### PR TITLE
ISPN-11345 Introduce integration tests for serverNG

### PR DESCRIPTION
**Pull request by diegolovison**
https://issues.redhat.com/browse/ISPN-11345

- [x] Support Arquillian
- [x] Arquillian is optional
- [x] Support multiple Infinispan Servers
- [X] Support Infinispan `CONTAINER` driver

**Support Arquillian**
Allow running the test in Wildfly or Tomcat

**Arquillian is optional**
We can run the test without deploying it in Wildfly or Tomcat. It runs faster.

**Support multiple Infinispan Servers**
As a QE, I would like to create a test scenario using multiple Infinispan Servers.

**Support Infinispan `CONTAINER` driver**
Infinispan Servers will be running in a container.

### CODE REVIEW

**Comment by tristantarrant** [comment]
Please wait while I work on extracting the Infinispan Server Test Driver to its own module

File: server/runtime/src/test/java/org/infinispan/server/test/InfinispanServerDriver.java [hunk 1]
```diff
@@ -97,7 +97,7 @@ public void prepare(String name) {
          throw new RuntimeException(e);
       }
       createUserFile("default");
-      createKeyStores();
+      //createKeyStores();
```
Comment (diegolovison) [inline server/runtime/src/test/java/org/infinispan/server/test/InfinispanServerDriver.java]: TODO

File: server/integration/src/test/java/org/infinispan/server/integration/HotRodRemoteCacheIT.java [hunk 2]
```diff
@@ -0,0 +1,1209 @@
...
+/**
+ * TODO.
+ *
+ * @author Richard Achmatowicz
+ * @author Martin Gencur
+ * @author Jozef Vilkolak
+ */
+public class HotRodRemoteCacheIT extends BaseIT {
+
+   @ClassRule
```
Comment (diegolovison) [inline server/integration/src/test/java/org/infinispan/server/integration/HotRodRemoteCacheIT.java]: @ClassRule
   public static ArquillianTestClass arquillianTestClass = new ArquillianTestClass();

   @Rule
   public ArquillianTest arquillianTest = new ArquillianTest();

**Comment by diegolovison** [comment]
@tristantarrant it is ready for review.

- I have added documentation about why not using a `TestRule` to start the Infinispan Server on the client-side.
- The most controversial are: 
-- `Class.forName("org.infinispan.server.integration.InstrumentArquillianContainer")` that is the way that I found to run the code only on the client-side.
-- `ClientSideOnlyDriver` to prevent adding a lot of dependencies in the container side

**Comment by diegolovison** [comment]
It is ready for review

File: server/integration/src/main/java/org/infinispan/server/integration/ArquillianServerType.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package org.infinispan.server.integration;
+
+public enum ArquillianServerType {
+
+   NONE, TOMCAT9, WILDFLY18;
```
Comment (tristantarrant) [inline server/integration/src/main/java/org/infinispan/server/integration/ArquillianServerType.java:5]: We should remove the version from the identifiers.

File: server/integration/pom.xml [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+
+    <artifactId>infinispan-server-integration</artifactId>
+    <name>Infinispan Server - Integration Tests</name>
+    <description>Infinispan Server - Integration Tests</description>
+
+    <properties>
+        <version.tomcat>9.0.11</version.tomcat>
+        <version.wildfly>18.0.1.Final</version.wildfly>
```
Comment (tristantarrant) [inline server/integration/pom.xml:17]: 19 please

File: server/integration/src/main/java/org/infinispan/server/integration/DeploymentBuilder.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+/**
+ * Responsible to build the war when needed
+ */
+public class DeploymentBuilder {
+
+   public static WebArchive war() {
+      WebArchive war = ShrinkWrap.create(WebArchive.class, "infinispan-server-integration.war");
+      addDefaultTestLib(war);
+      ArquillianServerType arquillianServerType = ArquillianServerType.current();
+      if (ArquillianServerType.TOMCAT9.equals(arquillianServerType)) {
```
Comment (tristantarrant) [inline server/integration/src/main/java/org/infinispan/server/integration/DeploymentBuilder.java:29]: Prefer a switch/case

File: server/integration/src/main/java/org/infinispan/server/integration/InfinispanServerIntegrationProcessorExecuter.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+import org.infinispan.server.test.core.ServerRunMode;
+import org.jboss.arquillian.core.api.annotation.Observes;
+import org.jboss.arquillian.test.spi.event.suite.AfterClass;
+import org.jboss.arquillian.test.spi.event.suite.BeforeClass;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+
+/*
+ * It starts the ServerRunMode.CONTAINER in the client side
+ */
+public class InfinispanServerIntegrationProcessorExecuter {
```
Comment (tristantarrant) [inline server/integration/src/main/java/org/infinispan/server/integration/InfinispanServerIntegrationProcessorExecuter.java:21]: s/Executer/Executor/

**Comment by diegolovison** [comment]
I am waiting for https://github.com/infinispan/infinispan/pull/8173 get merged and then will work in the master branch

**Comment by diegolovison** [comment]
I am waiting for https://github.com/infinispan/infinispan/pull/8297 get merged and then will work in the master branch

**Comment by danberindei** [comment]
Please rebase on master. You can also edit the PR and change the base to master, no need to open a new one.

**Comment by diegolovison** [comment]
I am going to follow https://github.com/infinispan/infinispan/tree/master/integrationtests/wildfly-modules that will simplify a lot the code

**Pull request closed without merging**